docs/source/developers/guide/step_by_step/testing.rst

```diff
@@ -62,29 +62,101 @@ In this section we outline steps needed for unit testing in Arrow.
+
+      **Structure**
+
+      Expect the usual testthat folder structure:
+
+      .. code-block:: R
+
+         tests
+          ├── testthat      # test files live here
+          └── testthat.R    # runs tests when R CMD check runs (e.g. with devtools::check())
+
+      This is the fundamental structure of testing in R with ``testthat``. Files such as ``testthat.R`` are not expected to change very often. For the ``arrow`` R package ``testthat.R`` also defines how the results of the various tests are displayed / reported in the console.
+
+      Usually, most files in the ``R/`` sub-folder have a corresponding test file in ``tests/testthat``.
+
+      **Running tests**
+
+      To run all tests in a package locally call
+
+      .. code-block:: R
+
+         devtools::test()
+
+      in the R console. Alternatively, you can use
+
+      .. code:: console
+
+         $ make test
+
+      in the shell.
+
+      You can run the tests in a single test file you have open with
+
+      .. code-block:: R
+
+         devtools::test_active_file()
+
+      All tests are also run as part of our continuous integration (CI) pipelines.
+
+      The Arrow R Developer guide also has a section on running tests. You can check it out `here <https://arrow.apache.org/docs/r/articles/developing.html#running-tests>`_.
+
+      **Good practice**
+
+      In general any change to source code needs to be accompanied by unit tests. All tests are expected to pass before a pull request is merged.
+
+      * Add functionality -> add unit tests
+      * Modify functionality -> update unit tests
+      * Solve a bug -> add unit test before solving it, which helps prove the bug and its fix
+      * Performance improvements should be reflected in benchmarks (which are also tests)
+      * An exception could be refactoring functionality that is fully covered by unit tests
+
+      If the new functionality is a user-facing or API change, you will almost certainly need to change tests — if no tests need to be changed it might mean the tests aren't right! If the new functionality is a refactor and no APIs are changing, there might not need to be test changes.
+
+      **Testing helpers**
+
+      To complement the ``testthat`` functionality, the ``arrow`` R package has defined a series of specific utility functions (called helpers), such as:
+
+      * expectations - these start with ``expect_`` and are used to compare objects
+            - for example, ``expect_altrep_roundtrip()`` compares the result of a function ``fn`` run on a vector ``x`` with the result of the same function run on the altrep version of ``x``. More generally, ``expect_…_roundtrip()`` functions take an input, convert it to some other format (e.g. arrow) and then convert it back, confirming that the values are the same.
```

Comment:
I wonder if we should also provide an example here that shows a call? The two function arguments feel a bit out of place / floating out there: either we could drop them, or we should show them in use in `expect_altrep_roundtrip()`

Reply:
Done. I changed the order of the sentences and went with general description of `expect_..._roundtrip()` + example